### PR TITLE
Validate that state-key is unique when using multiple callbacks of the same type

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

If you give two different callback instances with the same state key to the Trainer, they overwrite each other when saving a checkpoint. Only one of two states will actually be in the checkpoint. Hence, the trainer would fail to resume all states correctly (silently). 

This PR validates this at the time of the trainer instantiation. 

This change is required to continue with an improved version of #15606 to make checkpoint migration better.

### Does your PR introduce any breaking changes? If yes, please list them.

If someone was affected by this but weren't noticing it, when they upgrade Lighting they will get a friendly error now. Apart from that change, there are no intended breaking behaviors.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @awaelchli

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 1]
```diff
@@ -290,3 +294,18 @@ def _configure_external_callbacks() -> List[Callback]:
         )
         external_callbacks.extend(callbacks_list)
     return external_callbacks
+
+
+def _validate_callbacks_list(callbacks: List[Callback]) -> None:
+    stateful_callbacks = [cb for cb in callbacks if is_overridden("state_dict", instance=cb, parent=Callback)]
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:300]:
```suggestion
    stateful_callbacks = [cb for cb in callbacks if is_overridden("state_dict", cb)]
```

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 2]
```diff
@@ -290,3 +294,18 @@ def _configure_external_callbacks() -> List[Callback]:
...
+    seen_callbacks = defaultdict(list)
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:301]: Why a defaultdict instead of a plain old dict?

Why do we need a list? This enforces there's only 1 callback per state_key

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/callback_connector.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/callback_connector.py]: Oh, I don't need it anymore. Originally I wanted to print more information in the error message. We can just make it a set of state_keys.

File: src/pytorch_lightning/trainer/connectors/callback_connector.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/callback_connector.py:299]: I think we need the parent here. What if we have this inheritance: A > B > C where A is the super class of B and B is super class of C. If B implements state_dict, but C doesn't, I still want to categorize this as stateful.

Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/callback_connector.py:299]: Oh, you are thinking that `is_overridden` is smarter than it is and actually checks the inheritance tree. Not the case: https://github.com/Lightning-AI/lightning/blob/43fd9c1534411a80d5364031b3e529fef02a930c/src/pytorch_lightning/utilities/model_helpers.py#L31-L32

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/callback_connector.py:299]: Double oh, I was importing from lightning_utilities and this one isn't smart. Agree, I should be using the one from PL and then I can drop the parent! Great!

**Review by Borda — APPROVED** [review]
